### PR TITLE
Small tx response verification optimisation

### PR DESCRIPTION
**Pull request by emhane**
Replaces `Vec` for `HashSet` in `RequestTxHashes`. This simplifies the look up of a hash from a `PooledTransactions` response in `RequestHashes`, which is done in verifying that hashes in the response were actually requested.

### CODE REVIEW

File: crates/net/eth-wire/src/types/broadcast.rs [hunk 1]
```diff
@@ -687,30 +691,48 @@ pub struct RequestTxHashes {
...
+        if rest_capacity == 0 {
+            return Self::empty()
         }
+        let mut rest = Self::with_capacity(rest_capacity);
 
-        hashes.shrink_to_fit();
+        let mut i = 0;
+        self.hashes.retain(|hash| {
+            i += 1;
+            if i <= count {
```
Comment (mattsse) [inline crates/net/eth-wire/src/types/broadcast.rs:722]: can flip the order here to get rid of <= and skip +=1 if already > count

**Review by mattsse — COMMENTED** [review]
all of this looks reasonable,

though one bad side-effect of this is that we now request hashes in random order, which can result in more work during pool import because we might need to reorder sequences from the same account that we request out of nonce order, assuming they are announced in order

wdyt @Rjected

**Comment by emhane** [comment]
> all of this looks reasonable,
> 
> though one bad side-effect of this is that we now request hashes in random order, which can result in more work during pool import because we might need to reorder sequences from the same account that we request out of nonce order, assuming they are announced in order
> 
> wdyt @Rjected

random order already happens when the announcement is validated when we data is put into hash map to dedup

File: crates/net/eth-wire/src/types/broadcast.rs (hunk 1 above)
Comment (emhane) [reply crates/net/eth-wire/src/types/broadcast.rs:722]: count needs to increment every iteration

Comment (emhane) [reply crates/net/eth-wire/src/types/broadcast.rs]: ok you mean the true and false arms

File: crates/net/eth-wire/src/types/broadcast.rs [hunk 2]
```diff
@@ -868,4 +891,108 @@ mod tests {
...
+        assert_eq!(3, hashes.len());
+        assert_eq!(2, rest.len());
+    }
+
+    #[test]
+    fn request_hashes_retain_count_keep_all() {
+        let mut hashes = RequestTxHashes::new(
+            vec![
+                B256::from_str(
+                    "0x0000000000000000000000000000000000000000000000000000000000000001",
```
Comment (DaniPopes) [inline crates/net/eth-wire/src/types/broadcast.rs:935]: can use b256! macro

File: crates/net/eth-wire/src/types/broadcast.rs [hunk 3]
```diff
@@ -868,4 +891,108 @@ mod tests {
...
+        let rest = hashes.retain_count(3);
+
+        assert_eq!(3, hashes.len());
+        assert_eq!(2, rest.len());
+    }
+
+    #[test]
+    fn request_hashes_retain_count_keep_all() {
+        let mut hashes = RequestTxHashes::new(
+            vec![
```
Comment (DaniPopes) [inline crates/net/eth-wire/src/types/broadcast.rs:933]: doesn't need vec!

File: crates/net/eth-wire/src/types/broadcast.rs (hunk 2 above)
Comment (emhane) [reply crates/net/eth-wire/src/types/broadcast.rs:935]: ooo

File: crates/net/network/src/transactions/fetcher.rs [hunk 4]
```diff
@@ -160,7 +160,7 @@ impl TransactionFetcher {
                 }
             }
         };
-        let hash = hashes_to_request.first()?;
+        let hash = hashes_to_request.iter().next()?;
```
Comment (DaniPopes) [inline crates/net/network/src/transactions/fetcher.rs:163]: This won't be deterministic anymore, don't know if that matters here but it's something to note.

**Review by DaniPopes — APPROVED** [review]

File: crates/net/network/src/transactions/fetcher.rs (hunk 4 above)
Comment (emhane) [reply crates/net/network/src/transactions/fetcher.rs:163]: at this point `hashes_to_request` will only contain one hash

**Comment by emhane** [comment]
> > all of this looks reasonable,
> > though one bad side-effect of this is that we now request hashes in random order, which can result in more work during pool import because we might need to reorder sequences from the same account that we request out of nonce order, assuming they are announced in order
> > wdyt @Rjected
> 
> random order already happens when the announcement is validated when data is put into hash map to dedup

if you think this is a problem now (if I remember correct you wished for the `HashMap` in dedup in validation), reminder to open an issue since is out of scope @mattsse

**Pull request merged**